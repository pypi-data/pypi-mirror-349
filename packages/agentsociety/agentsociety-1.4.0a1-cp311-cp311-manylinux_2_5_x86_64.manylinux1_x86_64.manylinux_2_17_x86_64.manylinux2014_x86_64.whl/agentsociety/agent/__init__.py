from .agent import (
    BankAgentBase,
    CitizenAgentBase,
    FirmAgentBase,
    GovernmentAgentBase,
    NBSAgentBase,
)
from .agent_base import Agent, AgentToolbox, AgentType
from .block import Block, log_and_check, log_and_check_with_memory, trigger_class
from .prompt import FormatPrompt
from .trigger import EventTrigger, MemoryChangeTrigger, TimeTrigger

__all__ = [
    "Agent",
    "CitizenAgentBase",
    "AgentType",
    "AgentToolbox",
    "FirmAgentBase",
    "BankAgentBase",
    "NBSAgentBase",
    "GovernmentAgentBase",
    "MemoryChangeTrigger",
    "TimeTrigger",
    "EventTrigger",
    "Block",
    "log_and_check",
    "log_and_check_with_memory",
    "FormatPrompt",
    "trigger_class",
]
