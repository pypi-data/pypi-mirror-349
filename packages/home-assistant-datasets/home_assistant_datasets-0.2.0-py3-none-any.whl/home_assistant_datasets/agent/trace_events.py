"""The trace_events module is related to events captured when scraping model outputs.

This is used during the collect phase during the conversation agent calls to
capture runtime data or tracing such as details of the conversation or other
stats.

This can also be used in the evaluation state to interpret the collected
model outputs or tracing.
"""

import dataclasses
from dataclasses import dataclass
from typing import Any

from homeassistant.components.conversation import trace


__all__ = [
    "TokenStats",
    "TokenStatsBank",
    "find_llm_call",
    "find_token_stats",
]


@dataclass(frozen=True, kw_only=True)
class TokenStats:
    """Class for token stats."""

    input_tokens: int | float
    cached_input_tokens: int | float
    output_tokens: int | float
    duration_ms: float = 0
    n_count: int | float = 1
    """Total raw number of requests, which may be more than number of tasks."""


class TokenStatsBank:
    """Class for summarizing TokenStats across a bunch of runs."""

    def __init__(self) -> None:
        """Initialize report."""
        self.stats: list[TokenStats] = []

    def append(self, stats: TokenStats) -> None:
        """Append a token stats record."""
        self.stats.append(stats)

    def summary_data(self) -> dict[str, Any]:
        """Return a summary of the token stats as a dictionary."""
        return {
            "token_avg": dataclasses.asdict(self.avg()),
            "token_sum": dataclasses.asdict(self.sum()),
            "token_input_cache_ratio": round(
                sum(s.cached_input_tokens for s in self.stats)
                / sum(s.input_tokens for s in self.stats),
                2,
            ),
        }

    def avg(self) -> TokenStats:
        """Average the number of tokens across tasks."""
        return TokenStats(
            input_tokens=round(
                sum(s.input_tokens for s in self.stats) / len(self.stats), 2
            ),
            cached_input_tokens=round(
                sum(s.cached_input_tokens for s in self.stats) / len(self.stats), 2
            ),
            output_tokens=round(
                sum(s.output_tokens or 0 for s in self.stats) / len(self.stats), 2
            ),
            duration_ms=round(
                sum(s.duration_ms for s in self.stats) / len(self.stats), 2
            ),
            n_count=round(sum(s.n_count for s in self.stats) / len(self.stats), 2),
        )

    def sum(self) -> TokenStats:
        """Sum the token stats across tasks."""
        return TokenStats(
            input_tokens=sum(s.input_tokens for s in self.stats),
            cached_input_tokens=sum(s.cached_input_tokens for s in self.stats),
            output_tokens=sum(s.output_tokens or 0 for s in self.stats),
            duration_ms=sum(s.duration_ms for s in self.stats),
            n_count=sum(s.n_count for s in self.stats),
        )


def find_token_stats(
    trace_events: list[dict[str, Any]], duration_ms: float
) -> TokenStats | None:
    """Gets the agent detail that contains conversation agent statistics."""
    stats_data = list(
        stats
        for event in trace_events
        if event["event_type"] == trace.ConversationTraceEventType.AGENT_DETAIL
        and (stats := event["data"].get("stats")) is not None
    )
    if not stats_data:
        return None
    bank = TokenStatsBank()
    for stats in stats_data:
        bank.append(
            TokenStats(
                input_tokens=stats.get("input_tokens", 0),
                cached_input_tokens=stats.get("cached_input_tokens", 0),
                output_tokens=stats.get("output_tokens", 0),
                duration_ms=duration_ms,
            )
        )
        duration_ms = 0
    return bank.sum()


def find_llm_call(trace_events: list[dict[str, Any]]) -> dict[str, Any] | None:
    """Gets the llm call from the conversation trace."""
    tool_call = next(
        iter(
            event
            for event in trace_events
            if event["event_type"]
            # type: ignore[attr-defined]
            in (trace.ConversationTraceEventType.TOOL_CALL, "llm_tool_call")
        ),
        None,
    )
    if tool_call is None:
        return None

    data = tool_call["data"]
    return {
        "tool_name": data.get("tool_name"),
        "tool_args": data.get("tool_args"),
    }


def token_stats_from_context(context: dict[str, Any]) -> TokenStats | None:
    """Extract the TokenStats from the collection context."""
    return find_token_stats(
        context.get("conversation_trace", {}), duration_ms=context.get("duration_ms", 0)
    )
