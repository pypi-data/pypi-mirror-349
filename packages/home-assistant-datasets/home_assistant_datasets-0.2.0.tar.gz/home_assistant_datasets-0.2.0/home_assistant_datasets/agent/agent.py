"""Conversation agent module."""

from abc import ABC, abstractmethod
from typing import Any

from homeassistant.core import HomeAssistant


class ConversationAgent(ABC):
    """A simple client library for calling a conversation agent."""

    @abstractmethod
    async def async_process(self, hass: HomeAssistant, text: str) -> str:
        """Process a text input and return the response."""

    @abstractmethod
    def trace_context(self) -> dict[str, Any]:
        """Record any relevant trace context captured during the requests."""
