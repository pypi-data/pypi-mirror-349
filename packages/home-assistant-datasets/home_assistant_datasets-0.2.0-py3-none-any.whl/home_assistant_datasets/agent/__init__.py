"""Conversation agent module.

This module is a thin wrapper around the Conversation Agent process service,
tied to a simple interface to handle things like rate limiting, timing,
and retries.

This module is primarily used during data collection / scraping.
"""

from . import rate_limit, retryable, service_call, timing
from .agent import ConversationAgent


__all__ = [
    "ConversatioAgent",
    "create_default_agent",
    "trace_events",
    "service_call",
    "rate_limit",
    "retryable",
    "timing",
]


def create_default_agent(
    conversation_agent_id: str, rpm: int | None = None
) -> ConversationAgent:
    """Create the conversation agent client id.

    Note that this should be long lived in order for rate limiting to be effective.
    """
    agent = service_call.create_agent(conversation_agent_id)
    agent = timing.timed_agent(agent)
    if rpm:
        agent = rate_limit.wrap_rate_limit(agent, rpm)
    agent = retryable.wrap_retryable(agent)
    return agent
