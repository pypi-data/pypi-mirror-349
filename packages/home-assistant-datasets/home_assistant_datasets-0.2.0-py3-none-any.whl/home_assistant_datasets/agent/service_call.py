"""The default conversation agent implemented with a service call."""

import datetime
import logging
from typing import Any

from homeassistant.core import HomeAssistant, Context
from homeassistant.components.conversation import trace
from homeassistant.helpers import llm

from .agent import ConversationAgent


__all__ = [
    "create_agent",
]

_LOGGER = logging.getLogger(__name__)


def dump_conversation_trace(trace: trace.ConversationTrace) -> list[dict[str, Any]]:
    """Serialize the conversation trace for evaluation."""
    trace_data = trace.as_dict()
    trace_events = trace_data["events"]
    result = []
    for trace_event in trace_events:
        trace_event_data = trace_event["data"]
        data = {}
        for k, v in trace_event_data.items():
            if isinstance(v, Context):
                v = dict(v.as_dict())
            if isinstance(v, list) and v and isinstance(v[0], llm.Tool):
                v = [
                    {
                        "name": tool.name,
                        "description": tool.description,
                        "parameters": str(tool.parameters),
                    }
                    for tool in v
                ]
            data[k] = v
        values: dict[str, Any] = {
            "event_type": str(trace_event["event_type"]),
            "data": data,
        }
        if ts_str := trace_event.get("timestamp"):
            values["timestamp"] = datetime.datetime.fromisoformat(ts_str)
        result.append(values)
    return result


class ServiceCall(ConversationAgent):
    """A client library for a conversation agent service call."""

    def __init__(self, agent_id: str) -> None:
        """Initialize the agent."""
        self._agent_id = agent_id

    async def async_process(self, hass: HomeAssistant, text: str) -> str:
        """Process a text input and return the response."""
        _LOGGER.debug("hass.services.async_call=%s", hass.services.async_call)
        service_response = await hass.services.async_call(
            "conversation",
            "process",
            {"agent_id": self._agent_id, "text": text},
            blocking=True,
            return_response=True,
        )
        assert service_response
        response = service_response["response"]
        return str(response["speech"]["plain"]["speech"])  # type: ignore[call-overload, index]

    def trace_context(self) -> dict[str, Any]:
        """Record any relevant trace context captured during the requests."""
        conversation_trace = []
        if (traces := trace.async_get_traces()) and (last_trace := traces[-1]):
            conversation_trace = dump_conversation_trace(last_trace)
        return {
            "conversation_trace": conversation_trace,
        }


def create_agent(agent_id: str) -> ConversationAgent:
    """Create a conversation agent service call wrapper."""
    return ServiceCall(agent_id)
