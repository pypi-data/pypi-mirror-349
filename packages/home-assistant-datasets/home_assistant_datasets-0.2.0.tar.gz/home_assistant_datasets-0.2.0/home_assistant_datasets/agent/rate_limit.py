"""Rate limiter to avoid throttling from APIs."""

from typing import Any

from pyrate_limiter import (
    Duration,
    Rate,
    Limiter,
)

from homeassistant.core import HomeAssistant

from .agent import ConversationAgent

__all__ = [
    "wrap_rate_limit",
]

MAX_DELAY = 60 * 1000  # 1 minute


def _create_limiter(rpm: int) -> Limiter:
    """Create a rate limiter."""
    rate = Rate(rpm, Duration.MINUTE)
    return Limiter(rate, max_delay=MAX_DELAY)


class RateLimitedAgent(ConversationAgent):
    """A client library for a conversation agent service call."""

    def __init__(self, agent: ConversationAgent, rate_limiter: Limiter) -> None:
        """Initialize the agent."""
        self._agent = agent
        self._rate_limiter = rate_limiter

    async def async_process(self, hass: HomeAssistant, text: str) -> str:
        """Process a text input and return the response."""
        if self._rate_limiter:
            self._rate_limiter.try_acquire("item")
        return await self._agent.async_process(hass, text)

    def trace_context(self) -> dict[str, Any]:
        """Record any relevant trace context captured during the requests."""
        return self._agent.trace_context()


def wrap_rate_limit(agent: ConversationAgent, rpm: int) -> ConversationAgent:
    """Rate limit the specified agent.

    Note that this should be long lived in order for rate limiting to be effective.
    """
    return RateLimitedAgent(agent, _create_limiter(rpm))
