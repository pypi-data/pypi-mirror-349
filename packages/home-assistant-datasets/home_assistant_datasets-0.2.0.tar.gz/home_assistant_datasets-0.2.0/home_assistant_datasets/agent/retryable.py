"""A wrapper to make a conversation agent retryable."""

import asyncio
import logging
import json
from typing import Any

from homeassistant.core import HomeAssistant
from homeassistant.exceptions import HomeAssistantError, ServiceValidationError

from .agent import ConversationAgent

__all__ = [
    "wrap_retryable",
]


_LOGGER = logging.getLogger(__name__)


TIMEOUT = 40
MAX_TRIES = 3


class RetryableAgent(ConversationAgent):
    """A retryable conversation agent wrapper."""

    def __init__(self, agent: ConversationAgent) -> None:
        """Initialize the agent."""
        self._agent = agent
        self._tries = 0

    async def async_process(self, hass: HomeAssistant, text: str) -> str:
        """Process a text input and return the response."""
        # Run the conversation agent
        self._tries = 0
        response = ""
        retryable = True
        while self._tries < MAX_TRIES and retryable:
            self._tries += 1
            retryable = False
            _LOGGER.debug("Prompt: %s", text)
            try:
                async with asyncio.timeout(TIMEOUT):
                    response = await self._agent.async_process(hass, text)
            except (
                HomeAssistantError,
                ServiceValidationError,
                TypeError,
                json.JSONDecodeError,
            ) as err:
                response = str(err)
            except (TimeoutError, asyncio.CancelledError):
                _LOGGER.debug("Timeout error (tries=%s)", self._tries)
                response = f"Timeout (after {self._tries} tries)"
                retryable = True
            await hass.async_block_till_done()
            _LOGGER.debug("Response: %s", response)

        return response

    def trace_context(self) -> dict[str, Any]:
        """Record any relevant trace context captured during the requests."""
        context = self._agent.trace_context()
        context["tries"] = self._tries
        return context


def wrap_retryable(agent: ConversationAgent) -> ConversationAgent:
    """Wrap the agent with retry logic, returning errors as text."""
    return RetryableAgent(agent)
