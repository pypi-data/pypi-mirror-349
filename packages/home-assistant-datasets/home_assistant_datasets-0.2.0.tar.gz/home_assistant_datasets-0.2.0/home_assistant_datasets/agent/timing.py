"""A wrapper to make a conversation agent retryable."""

import datetime
import logging
from typing import Any

from homeassistant.core import HomeAssistant

from .agent import ConversationAgent


__all__ = [
    "timed_agent",
]


_LOGGER = logging.getLogger(__name__)


class TimedAgent(ConversationAgent):
    """A conversation agent wrapper that tracks called durations."""

    def __init__(self, agent: ConversationAgent) -> None:
        """Initialize the agent."""
        self._agent = agent
        self._elapsed: datetime.timedelta | None = None

    async def async_process(self, hass: HomeAssistant, text: str) -> str:
        """Process a text input and return the response."""
        start = datetime.datetime.now()
        try:
            return await self._agent.async_process(hass, text)
        finally:
            self._elapsed = datetime.datetime.now() - start

    def trace_context(self) -> dict[str, Any]:
        """Record any relevant trace context captured during the requests."""
        context = self._agent.trace_context()
        if self._elapsed is not None:
            duration_ms = self._elapsed / datetime.timedelta(milliseconds=1)
            context["duration_ms"] = duration_ms
        return context


def timed_agent(agent: ConversationAgent) -> ConversationAgent:
    """Factory function that wraps the conversation agent with the timer."""
    return TimedAgent(agent)
