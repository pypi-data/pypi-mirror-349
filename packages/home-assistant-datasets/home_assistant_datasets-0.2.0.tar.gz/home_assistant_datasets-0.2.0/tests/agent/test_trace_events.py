"""Tests for the trace_events module."""

from syrupy import SnapshotAssertion

from home_assistant_datasets.agent.trace_events import TokenStats, TokenStatsBank


def test_token_stats_bank(snapshot: SnapshotAssertion) -> None:
    """Tests for TokenStatsBank."""

    bank = TokenStatsBank()
    bank.append(
        TokenStats(
            input_tokens=500,
            cached_input_tokens=100,
            output_tokens=250,
            duration_ms=150.0,
        )
    )
    bank.append(
        TokenStats(
            input_tokens=500,
            cached_input_tokens=100,
            output_tokens=250,
            duration_ms=105.0,
        )
    )
    assert bank.summary_data() == snapshot
