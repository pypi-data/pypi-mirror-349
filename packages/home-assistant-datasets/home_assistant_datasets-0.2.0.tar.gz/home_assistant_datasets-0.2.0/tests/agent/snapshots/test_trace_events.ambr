# serializer version: 1
# name: test_token_stats_bank
  dict({
    'token_avg': dict({
      'cached_input_tokens': 100.0,
      'duration_ms': 127.5,
      'input_tokens': 500.0,
      'n_count': 1.0,
      'output_tokens': 250.0,
    }),
    'token_input_cache_ratio': 0.2,
    'token_sum': dict({
      'cached_input_tokens': 200,
      'duration_ms': 255.0,
      'input_tokens': 1000,
      'n_count': 2,
      'output_tokens': 500,
    }),
  })
# ---
